torch/csrc/distributed/rpc/future_message.cpp

```diff
@@ -16,14 +19,28 @@ const Message& FutureMessage::wait() {
   return message_;
 }
 
+const Message& FutureMessage::waitNoThrow() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  finished_cv_.wait(lock, [this] { return completed_.load(); });
+  return message_;
+}
+
+Message&& FutureMessage::moveMessage() && {
+  std::unique_lock<std::mutex> lock(mutex_);
+  return std::move(message_);
+}
+
 void FutureMessage::markCompleted(Message message) {
   {
     std::unique_lock<std::mutex> lock(mutex_);
     TORCH_CHECK(!completed());
-    completed_ = true;
     message_ = std::move(message);
-
-    fireCallbacks();
+    completed_ = true;
+    std::vector<Callback> cbs = std::move(callbacks_);
+    lock.unlock();
+    for (auto& callback : cbs) {
+      callback(message_);
+    }
```

Comment:
callbacks_.clear()?

Reply:
callbacks_.clear() is implicit here because of the std::move() from the vector.
That said, I'll replace with a swap with an empty vector, which should be more explicit.